### PR TITLE
[MXNet-1349][Fit API]Add validation support and unit tests for fit() API

### PR DESCRIPTION
**Pull request by abhinavs95**
## Description ##
Adding validation support and unit tests for fit() API.
This PR depends on the parent PR for fit() API #14346 
JIRA epic: https://issues.apache.org/jira/projects/MXNET/issues/MXNET-1333
Design: https://cwiki.apache.org/confluence/display/MXNET/Gluon+Fit+API+-+Tech+Design

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNet-1349](https://issues.apache.org/jira/browse/MXNET-1349)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
1. Added validation support in estimator class, allow users to monitor loss and metrics on validation set during training.
2. Added unit tests for estimator and fit()
3. Added error handlers in estimator

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: python/mxnet/gluon/estimator/estimator.py [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+from ... import gluon, autograd
+from ...context import Context, cpu, gpu, num_gpus
+from ...io import DataIter
+from ...metric import EvalMetric, Loss
+import copy
+
+__all__ = ['Estimator']
+
+
+class Estimator(object):
```
Comment (piyushghai) [inline python/mxnet/gluon/estimator/estimator.py]: Wasn't this class included in this PR : https://github.com/apache/incubator-mxnet/pull/14346 ?

Comment (abhinavs95) [reply python/mxnet/gluon/estimator/estimator.py]: Yes, I have made changes in it for validation support.

Comment (piyushghai) [reply python/mxnet/gluon/estimator/estimator.py]: I'd recommend either adding the changes for validation metrics as a commit to the existing PR #14346 , OR, wait until that PR gets merged, and add this as a singular commit so that the diff is incremental.

File: python/mxnet/gluon/estimator/estimator.py [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+            data = batch.data[0]
+            label = batch.label[0]
+        else:
+            data = batch[0]
+            label = batch[1]
+        data = gluon.utils.split_and_load(data, ctx_list=ctx, batch_axis=0)
+        label = gluon.utils.split_and_load(label, ctx_list=ctx, batch_axis=0)
+        return data, label
+
+    def _test(self, val_data, batch_fn=None):
```
Comment (roywei) [inline python/mxnet/gluon/estimator/estimator.py]: maybe rename to `_evaluate(self, eval_data)`, as this can be used for both validation (on epoch end) and test(on train end) dataset.

File: python/mxnet/gluon/estimator/estimator.py [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+        for metric in self.test_metrics:
+            self.train_stats['test_' + metric.name] = []
+        self.train_loss_metrics = []
+        self.test_loss_metrics = []
+        # using the metric wrapper for loss to record loss value
+        for l in self.loss:
+            self.train_loss_metrics.append(Loss(l.name))
+            self.test_loss_metrics.append(Loss(l.name))
+            self.train_stats['train_' + l.name] = []
+            self.train_stats['test_' + l.name] = []
```
Comment (roywei) [inline python/mxnet/gluon/estimator/estimator.py]: use `val_`

File: python/mxnet/gluon/estimator/estimator.py [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+        self.train_stats['learning_rate'] = []
+        # current step of the epoch
+        self.train_stats['step'] = ''
+        for metric in self.train_metrics:
+            # record a history of metrics over each epoch
+            self.train_stats['train_' + metric.name] = []
+            # only record the latest metric numbers after each batch
+            self.train_stats['batch_' + metric.name] = 0.
+        for metric in self.test_metrics:
+            self.train_stats['test_' + metric.name] = []
```
Comment (roywei) [inline python/mxnet/gluon/estimator/estimator.py]: use `val_`

File: python/mxnet/gluon/estimator/event_handler.py [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+    def epoch_begin(self):
+        self.epoch_start = time.time()
+
+    def epoch_end(self, do_validation=False):
+        epoch_time = time.time() - self.epoch_start
+        epoch = self._estimator.train_stats['epochs'][-1]
+        msg = '\n[Epoch %d] finished in %.3fs: ' % (epoch, epoch_time)
+        for key in self._estimator.train_stats.keys():
+            if do_validation:
+                if key.startswith('train_') or key.startswith('test_'):
```
Comment (roywei) [inline python/mxnet/gluon/estimator/event_handler.py]: you can remove the if else and no need to pass `do_validation`.  The logic can simply be if key starts with `train` or `val`, log it, even if no  validation is done

File: python/mxnet/gluon/estimator/estimator.py (hunk 1 above)
Comment (roywei) [reply python/mxnet/gluon/estimator/estimator.py]: @piyushghai I asked @abhinavs95 to open this PR so I can review it early. We can do a rebase and resolve the conflict once parent PR is merged.

**Comment by karan6181** [comment]
@mxnet-label-bot add [Gluon, Test]

File: python/mxnet/gluon/estimator/estimator.py (hunk 2 above)
Comment (abhinavs95) [reply python/mxnet/gluon/estimator/estimator.py]: Done!

File: python/mxnet/gluon/estimator/estimator.py (hunk 3 above)
Comment (abhinavs95) [reply python/mxnet/gluon/estimator/estimator.py]: Done!

File: python/mxnet/gluon/estimator/estimator.py (hunk 4 above)
Comment (abhinavs95) [reply python/mxnet/gluon/estimator/estimator.py]: Done!

File: python/mxnet/gluon/estimator/event_handler.py (hunk 5 above)
Comment (abhinavs95) [reply python/mxnet/gluon/estimator/event_handler.py]: Done! Also moved val stats update loop to fit() to avoid key error. If no val set is passed 'nan' will be logged for val stats.

File: python/mxnet/gluon/estimator/estimator.py [hunk 6]
```diff
@@ -26,6 +26,7 @@
 from ...context import Context, cpu, gpu, num_gpus
 from ...io import DataIter
 from ...metric import EvalMetric, Loss
+import copy
```
Comment (roywei) [inline python/mxnet/gluon/estimator/estimator.py]: this failed sanity check: import it first

File: python/mxnet/gluon/estimator/estimator.py [hunk 7]
```diff
@@ -253,8 +294,14 @@ def fit(self, train_data,
                 for handler in event_handlers:
                     handler.batch_end()
 
-            for metric in self.metrics + self.loss_metrics:
+            if do_validation:
+                self._evaluate(val_data, batch_fn)
+
+            for metric in self.train_metrics + self.train_loss_metrics:
                 self.train_stats['train_' + metric.name].append(metric.get()[1])
+            for metric in self.test_metrics + self.test_loss_metrics:
```
Comment (piyushghai) [inline python/mxnet/gluon/estimator/estimator.py]: Can we rename ```test_metrics``` to ```val_metrics``` for consistency, since we are referring to them as validation stuff throughout

File: tests/python/unittest/test_gluon_estimator.py [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+    train_data = gluon.data.DataLoader(dataset, batch_size=batch_size)
+    loss = gluon.loss.L2Loss()
+    net.initialize(ctx=ctx)
+    trainer = gluon.Trainer(net.collect_params(), 'sgd', {'learning_rate': 0.001})
+    # input no metric
+    est = estimator.Estimator(net=net,
+                              loss=loss,
+                              trainers=trainer,
+                              context=ctx)
+    est.fit(train_data=train_data,
```
Comment (piyushghai) [inline tests/python/unittest/test_gluon_estimator.py]: What are we asserting against here ?

File: python/mxnet/gluon/estimator/estimator.py (hunk 7 above)
Comment (abhinavs95) [reply python/mxnet/gluon/estimator/estimator.py]: we can use `test_metrics` for both validation (at epoch end) and test (at train end) dataset

File: python/mxnet/gluon/estimator/estimator.py [hunk 9]
```diff
@@ -64,17 +65,21 @@ def __init__(self, net,
             self.loss = [loss]
         else:
             self.loss = loss or []
+            if not self.loss:
```
[the remark this answers is not in the document]
Comment (abhinavs95) [reply python/mxnet/gluon/estimator/estimator.py]: Done!

File: python/mxnet/gluon/estimator/estimator.py [hunk 10]
```diff
@@ -64,17 +65,21 @@ def __init__(self, net,
...
+            self.train_metrics = [metrics]
         else:
-            self.metrics = metrics or []
-            for metric in self.metrics:
+            self.train_metrics = metrics or []
+            for metric in self.train_metrics:
                 if not isinstance(metric, EvalMetric):
                     raise ValueError("metrics must be a Metric or a list of Metric, refer to mxnet.metric.EvalMetric")
+        # Use same metrics for validation
+        self.test_metrics = copy.deepcopy(self.train_metrics)
```
[the remark this answers is not in the document]
Comment (abhinavs95) [reply python/mxnet/gluon/estimator/estimator.py]: Done!

File: python/mxnet/gluon/estimator/estimator.py [hunk 11]
```diff
@@ -64,17 +65,21 @@ def __init__(self, net,
...
         else:
             self.loss = loss or []
+            if not self.loss:
+                raise ValueError("No loss specified, refer to gluon.loss.Loss")
             for l in self.loss:
-                if not isinstance(loss, gluon.loss.Loss):
+                if not isinstance(l, gluon.loss.Loss):
                     raise ValueError("loss must be a Loss or a list of Loss, refer to gluon.loss.Loss")
 
         if isinstance(metrics, EvalMetric):
```
[the remark this answers is not in the document]
Comment (abhinavs95) [reply python/mxnet/gluon/estimator/estimator.py]: Done!

File: python/mxnet/gluon/estimator/estimator.py [hunk 12]
```diff
@@ -156,7 +168,33 @@ def _batch_fn(self, batch, ctx, is_iterator=False):
         label = gluon.utils.split_and_load(label, ctx_list=ctx, batch_axis=0)
         return data, label
 
+    def _evaluate(self, val_data, batch_fn=None):
```
[the remark this answers is not in the document]
Comment (abhinavs95) [reply python/mxnet/gluon/estimator/estimator.py]: Can be done by wrapping X, y in a dataloader/dataiter and passing it as val_data

File: python/mxnet/gluon/estimator/estimator.py [hunk 13]
```diff
@@ -156,7 +168,33 @@ def _batch_fn(self, batch, ctx, is_iterator=False):
...
+        for metric in self.test_metrics + self.test_loss_metrics:
+            metric.reset()
+
+        for _, batch in enumerate(val_data):
+            if not batch_fn:
+                if isinstance(val_data, gluon.data.DataLoader):
```
[the remark this answers is not in the document]
Comment (abhinavs95) [reply python/mxnet/gluon/estimator/estimator.py]: this check needs to be before calling `batch_fn` as `val_data` is not available to it

File: python/mxnet/gluon/estimator/estimator.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (abhinavs95) [reply python/mxnet/gluon/estimator/estimator.py]: Done!

File: python/mxnet/gluon/estimator/estimator.py [hunk 14]
```diff
@@ -192,6 +230,9 @@ def fit(self, train_data,
                 not any(isinstance(handler, LoggingHandler) for handler in event_handlers):
             event_handlers.append(LoggingHandler(self))
 
+        # Check for validation data
+        do_validation = True if val_data else False
```
[the remark this answers is not in the document]
Comment (abhinavs95) [reply python/mxnet/gluon/estimator/estimator.py]: Done!

File: python/mxnet/gluon/estimator/estimator.py [hunk 15]
```diff
@@ -156,7 +168,33 @@ def _batch_fn(self, batch, ctx, is_iterator=False):
...
+            for loss in self.loss:
+                losses.append([loss(y_hat, y) for y_hat, y in zip(pred, label)])
+            # update metrics
+            for metric in self.test_metrics:
+                metric.update(label, pred)
+            for loss, loss_metric, in zip(losses, self.test_loss_metrics):
+                loss_metric.update(0, [l for l in loss])
+
     def fit(self, train_data,
+            val_data=None,
```
[the remark this answers is not in the document]
Comment (abhinavs95) [reply python/mxnet/gluon/estimator/estimator.py]: users might want to train without a validation set

File: tests/python/unittest/test_gluon_estimator.py (hunk 8 above)
Comment (abhinavs95) [reply tests/python/unittest/test_gluon_estimator.py]: this checks if the estimator works with no metric specified, doesn't throw any error/warning when its successful

File: python/mxnet/gluon/estimator/estimator.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (nswamy) [reply python/mxnet/gluon/estimator/estimator.py]: I don't see a reason why users would not a validation dataset, its required to know that the model is not overfitting/

File: python/mxnet/gluon/estimator/estimator.py [hunk 16]
```diff
@@ -62,37 +63,45 @@ def __init__(self, net,
...
-            self.loss = loss or []
-            for l in self.loss:
-                if not isinstance(loss, gluon.loss.Loss):
-                    raise ValueError("loss must be a Loss or a list of Loss, refer to gluon.loss.Loss")
+            raise ValueError("loss must be a Loss or a list of Loss, "
+                             "refer to gluon.loss.Loss:{}".format(loss))
 
         if isinstance(metrics, EvalMetric):
-            self.metrics = [metrics]
+            self.train_metrics = [metrics]
```
Comment (nswamy) [inline python/mxnet/gluon/estimator/estimator.py]: use 'Accuracy' as default when not passed.

File: python/mxnet/gluon/estimator/estimator.py [hunk 17]
```diff
@@ -62,37 +63,45 @@ def __init__(self, net,
...
         # current step of the epoch
         self.train_stats['step'] = ''
-        for metric in self.metrics:
+        for metric in self.train_metrics:
             # record a history of metrics over each epoch
             self.train_stats['train_' + metric.name] = []
             # only record the latest metric numbers after each batch
             self.train_stats['batch_' + metric.name] = 0.
-        self.loss_metrics = []
+        for metric in self.val_metrics:
```
Comment (karan6181) [inline python/mxnet/gluon/estimator/estimator.py]: Can we have one for loop for `self.train_metrics` and `self.val_metrics` since value for both the parameter is same according to line 82. Something like this `for train_m, val_m in zip(self.train_metrics, self.val_metrics)`. Though `zip()` operator stop after exhausting shorter array but since both the array are of same length, we can use `zip()` operator.

File: python/mxnet/gluon/estimator/estimator.py [hunk 18]
```diff
@@ -156,7 +168,46 @@ def _batch_fn(self, batch, ctx, is_iterator=False):
...
+            metric.reset()
+
+        for _, batch in enumerate(val_data):
+            if not batch_fn:
+                if isinstance(val_data, gluon.data.DataLoader):
+                    data, label = self._batch_fn(batch, self.context)
+                elif isinstance(val_data, DataIter):
+                    data, label = self._batch_fn(batch, self.context, is_iterator=True)
+                else:
+                    raise ValueError("You are using a custom iteration, please also provide "
```
Comment (karan6181) [inline python/mxnet/gluon/estimator/estimator.py]: It would be helpful to the end user if could provide more detailed exception. you append the below statement at the end,  something like this: `or you can provide the data in terms of gluon.data.DataLoader or mx.io.DataIter.`

File: python/mxnet/gluon/estimator/estimator.py (hunk 17 above)
Comment (abhinavs95) [reply python/mxnet/gluon/estimator/estimator.py]: We want to keep the train and val metrics separate here. Currently we are using the same metrics for val and train but future updates may involve separate user specified val metrics in which case combining this update loop won't work.

File: python/mxnet/gluon/estimator/estimator.py (hunk 18 above)
Comment (abhinavs95) [reply python/mxnet/gluon/estimator/estimator.py]: Thanks for pointing it out, updated!

File: python/mxnet/gluon/estimator/estimator.py (hunk 16 above)
Comment (abhinavs95) [reply python/mxnet/gluon/estimator/estimator.py]: 'Accuracy' will only work for classification cases, for other cases it will give inaccurate resutls or even fail. Also I'm not sure how we can infer metrics from loss function as there isn't a direct correlation between them, do you have any suggestions?

File: tests/python/unittest/test_gluon_estimator.py (hunk 8 above)
Comment (piyushghai) [reply tests/python/unittest/test_gluon_estimator.py]: So there's a particular happy path that you're trying to test here ?

File: python/mxnet/gluon/estimator/estimator.py (hunk 16 above)
Comment (nswamy) [reply python/mxnet/gluon/estimator/estimator.py]: I think we should still infer metrics from known Loss functions (at least from the examples you know)

Comment (abhinavs95) [reply python/mxnet/gluon/estimator/estimator.py]: Added Accuracy metric as default for SoftmaxCrossEntropy loss for now. Will add more in a followup PR.

File: python/mxnet/gluon/estimator/estimator.py [hunk 19]
```diff
@@ -77,6 +77,10 @@ def __init__(self, net,
                 raise ValueError("metrics must be a Metric or a list of Metric, "
                                  "refer to mxnet.metric.EvalMetric:{}".format(metrics))
 
+        # Use default mx.metric.Accuracy() for gluon.loss.SoftmaxCrossEntropyLoss()
+        if not self.train_metrics and any([isinstance(l, gluon.loss.SoftmaxCrossEntropyLoss) for l in self.loss]):
```
Comment (nswamy) [inline python/mxnet/gluon/estimator/estimator.py]: lets get this from a map of Loss->[default metrics] in the next version.

Comment (abhinavs95) [reply python/mxnet/gluon/estimator/estimator.py]: Yes, tracking it using JIRA issue: https://issues.apache.org/jira/browse/MXNET-1364
